server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+package org.http4s.server.middleware
+
+import java.nio.charset.StandardCharsets
+import java.security.{MessageDigest, SecureRandom}
+import java.time.Clock
+import javax.crypto.{KeyGenerator, Mac, SecretKey}
+import javax.crypto.spec.SecretKeySpec
+import java.util.Base64
+
+import org.http4s.{MaybeResponse, Request, Response, Status, Cookie}
+import org.http4s.server.Middleware
+import org.http4s.headers.{Cookie => HCookie}
+import org.http4s.util.{CaseInsensitiveString, NonEmptyList}
+
+import scalaz.{Kleisli, OptionT}
+import scalaz.concurrent.Task
+import scalaz.syntax.std.boolean._
+
+case class CSRFMiddleware(config: CSRFConfiguration,
+  clock: Clock = Clock.systemUTC()) {
+
+  /** Sign our token using the current time in milliseconds as a nonce
+    * Signing and generating a token is potentially a unsafe operation
+    * if constructed with a bad key.
+    */
+  protected[middleware] def signToken(token: String): Task[String] = {
+    val joined = token + "-" + clock.millis()
+    for {
+      instance <- Task.delay(Mac.getInstance(CSRFMiddleware.SigningAlgo))
+      _ <- Task.delay(instance.init(config.key))
+      out <- Task.delay(instance.doFinal(joined.getBytes("UTF-8")))
+    } yield joined + "-" + Base64.getEncoder.encodeToString(out)
```

Comment:
What does doing this monadically achieve vs. putting it in a single `Task.delay`?